src/circuit/gadget/sinsemilla/note_commit.rs

```diff
@@ -0,0 +1,1499 @@
+            let gd_x = meta.query_advice(config.advices[2], Rotation::next());
+            let pkd_x = meta.query_advice(config.advices[3], Rotation::next());
+            let value = meta.query_advice(config.advices[4], Rotation::next());
+            // b_0 has been constrained to be 4 bits outside this gate.
+            let b_0 = meta.query_advice(config.advices[5], Rotation::next());
+            // This gate constrains b_1 to be boolean.
+            let b_1 = meta.query_advice(config.advices[6], Rotation::next());
+            // This gate constrains d_0 to be boolean.
+            let d_0 = meta.query_advice(config.advices[7], Rotation::next());
+            // This gate constrains g_0 to be boolean.
+            let g_0 = meta.query_advice(config.advices[8], Rotation::next());
+            // z1_g has been constrained to 240 bits by the Sinsemilla hash.
+            let z1_g = meta.query_advice(config.advices[9], Rotation::next());
+            let g_2 = z1_g;
+
+            // Boolean checks on 1-bit pieces.
+            let boolean_checks = std::iter::empty()
+                .chain(Some(("bool_check b_1", bool_check(b_1.clone()))))
+                .chain(Some(("bool_check b_2", bool_check(b_2.clone()))))
+                .chain(Some(("bool_check d_0", bool_check(d_0.clone()))))
+                .chain(Some(("bool_check d_1", bool_check(d_1.clone()))))
+                .chain(Some(("bool_check g_0", bool_check(g_0.clone()))))
+                .chain(Some(("bool_check h_1", bool_check(h_1.clone()))));
+
+            // b = b_0 + (2^4) b_1 + (2^5) b_2 + (2^6) b_3
+            let b_check = b_whole
+                - (b_0.clone()
+                    + b_1.clone() * two_pow_4
+                    + b_2 * two_pow_5
+                    + b_3.clone() * two_pow_6);
+            // d = d_0 + (2) d_1 + (2^2) d_2 + (2^10) d_3
+            let d_check = d_whole
+                - (d_0.clone() + d_1 * two + d_2.clone() * two_pow_2 + d_3.clone() * two_pow_10);
+            // e = e_0 + (2^6) e_1
+            let e_check = e_whole - (e_0.clone() + e_1.clone() * two_pow_6);
+            // g = g_0 + (2) g_1 + (2^10) g_2
+            let g_check = g_whole - (g_0 + g_1.clone() * two + g_2.clone() * two_pow_10);
+            // h = h_0 + (2^5) h_1
+            let h_check = h_whole - (h_0 + h_1 * two_pow_5);
+
+            // Check that *_prime pieces were correctly derived.
+            // a_prime = a + 2^130 - t_P
+            let a_prime_check = a.clone() + two_pow_130 - t_p.clone() - a_prime;
+
+            // b3_c_prime = b_3 + (2^4)c + 2^140 - t_P
+            let b3_c_prime_check = b_3.clone() + (c.clone() * two_pow_4) + two_pow_140.clone()
+                - t_p.clone()
+                - b3_c_prime;
+
+            // e1_f_prime = e_1 + (2^4)f + 2^140 - t_P
+            let e1_f_prime_check =
+                e_1 + (f * two_pow_4) + two_pow_140.clone() - t_p.clone() - e1_f_prime;
+
+            // g1_g2_prime = g_1 + (2^9)g_2 + 2^140 - t_P
```

Comment:
This can actually be:
```suggestion
            // g1_g2_prime = g_1 + (2^9)g_2 + 2^130 - t_P
```

Unlike the other shifted decompositions (which use multiple message pieces), here we are using a subset of a single message piece, with `g_0` being a 1-bit chunk. Thus constraining `z13_g = 0` means that we've constrained `g_1 + (2^9)g_2 < 2^129 < 2^130`, so we can shorted the `g1_g2_prime` decomposition from 14 words to 13.